Linaria webpack loader can't resolve webpack alias

**Do you want to request a *feature* or report a *bug*?**

bug
**What is the current behavior?**

webpack config:
```
  resolve: {
    extensions: ['.js', '.jsx'],
    alias: {
      '@': `src/shared`,
    },
  },
```

file with linaria import:
```
import { css } from 'linaria';
import bp from '@/lib/breakpoints';
```

error:
```
Error: Module build failed (from ../node_modules/linaria/loader.js):
 SyntaxError: src/shared/components/TitleWithIcon/TitleWithIcon.styles.js: An error occurred when evaluating the expression: Cannot find module '@/lib/breakpoints'. Make sure you are not using a browser or Node specific API.
```

--- a/package.json
+++ b/package.json
@@ -86,6 +86,7 @@
     "@babel/register": "^7.0.0",
     "cosmiconfig": "^5.0.7",
     "dedent": "^0.7.0",
+    "enhanced-resolve": "^4.1.0",
     "glob": "^7.1.3",
     "loader-utils": "^1.1.0",
     "mkdirp": "^0.5.1",

--- a/src/__tests__/module.test.js
+++ b/src/__tests__/module.test.js
@@ -191,3 +191,23 @@ it('has __dirname available', () => {
 
   expect(mod.exports).toBe(path.dirname(mod.filename));
 });
+
+it('changes resolve behaviour on overriding _resolveFilename', () => {
+  const originalResolveFilename = Module._resolveFilename;
+
+  Module._resolveFilename = id => (id === 'foo' ? 'bar' : id);
+
+  const mod = new Module(path.resolve(__dirname, '../__fixtures__/test.js'));
+
+  mod.evaluate(dedent`
+  module.exports = [
+    require.resolve('foo'),
+    require.resolve('test'),
+  ];
+  `);
+
+  // Restore old behavior
+  Module._resolveFilename = originalResolveFilename;
+
+  expect(mod.exports).toEqual(['bar', 'test']);
+});

--- a/src/babel/module.js
+++ b/src/babel/module.js
@@ -25,11 +25,16 @@ const NOOP = () => {};
 class Module {
   static invalidate: () => void;
 
+  static _resolveFilename: (
+    id: string,
+    options: { id: string, filename: string, paths: string[] }
+  ) => string;
+
   id: string;
 
   filename: string;
 
-  paths: string;
+  paths: string[];
 
   require: (id: string) => any;
 
@@ -85,8 +90,7 @@ class Module {
         added.push(ext);
       });
 
-      // Resolve the module using node's resolve algorithm
-      return NativeModule._resolveFilename(id, this);
+      return Module._resolveFilename(id, this);
     } finally {
       // Cleanup the extensions we added to restore previous behaviour
       added.forEach(ext => delete extensions[ext]);
@@ -166,5 +170,11 @@ class Module {
 Module.invalidate = () => {
   cache = {};
 };
+
+// Alias to resolve the module using node's resolve algorithm
+// This static property can be overriden by the webpack loader
+// This allows us to use webpack's module resolution algorithm
+Module._resolveFilename = (id, options) =>
+  NativeModule._resolveFilename(id, options);
 
 module.exports = Module;

--- a/src/loader.js
+++ b/src/loader.js
@@ -4,9 +4,9 @@ const fs = require('fs');
 const path = require('path');
 const mkdirp = require('mkdirp');
 const normalize = require('normalize-path');
-/* $FlowFixMe */
-const Module = require('module');
 const loaderUtils = require('loader-utils');
+const enhancedResolve = require('enhanced-resolve/lib/node');
+const Module = require('./babel/module');
 const transform = require('./transform');
 
 module.exports = function loader(content: string, inputSourceMap: ?Object) {
@@ -27,13 +27,35 @@ module.exports = function loader(content: string, inputSourceMap: ?Object) {
     )
   );
 
-  const result = transform(content, {
-    filename: this.resourcePath,
-    inputSourceMap: inputSourceMap != null ? inputSourceMap : undefined,
-    outputFilename,
-    pluginOptions: rest,
-    preprocessor,
-  });
+  const resolveSync = enhancedResolve.create.sync(
+    // this._compilation is a deprecated API
+    // However there seems to be no other way to access webpack's resolver
+    // There is this.resolve, but it's asynchronous
+    // Another option is to read the webpack.config.js, but it won't work for programmatic usage
+    // This API is used by many loaders/plugins, so hope we're safe for a while
+    this._compilation.options.resolve
+  );
+
+  let result;
+
+  const originalResolveFilename = Module._resolveFilename;
+
+  try {
+    // Use webpack's resolution when evaluating modules
+    Module._resolveFilename = (id, { filename }) =>
+      resolveSync(path.dirname(filename), id);
+
+    result = transform(content, {
+      filename: this.resourcePath,
+      inputSourceMap: inputSourceMap != null ? inputSourceMap : undefined,
+      outputFilename,
+      pluginOptions: rest,
+      preprocessor,
+    });
+  } finally {
+    // Restore original behaviour
+    Module._resolveFilename = originalResolveFilename;
+  }
 
   if (result.cssText) {
     let { cssText } = result;
@@ -47,15 +69,11 @@ module.exports = function loader(content: string, inputSourceMap: ?Object) {
     if (result.dependencies && result.dependencies.length) {
       result.dependencies.forEach(dep => {
         try {
-          const f = Module._resolveFilename(dep, {
-            id: this.resourcePath,
-            filename: this.resourcePath,
-            paths: Module._nodeModulePaths(path.dirname(this.resourcePath)),
-          });
+          const f = resolveSync(path.dirname(this.resourcePath), dep);
 
           this.addDependency(f);
         } catch (e) {
-          // Ignore
+          console.warn(`[linaria] failed to add dependency for: ${dep}`, e);
         }
       });
     }

